Add a Unified Build design for full VMR code flow

premun opened the pull request:
The design still has 2 TODO sections and it doesn't contain the plan for the rollout of the new dependency flow but we agreed to plan that later once we know the backflow through and through.

Opening this for a discussion for the UB group.

https://github.com/dotnet/arcade-services/issues/2714

premun commented:
@jkotas I went back and forth on the xcopy idea and it seemed to be great but when I put the code flows within a context of a real world (aka there would be always a flow before the current one), the xcopy started to break apart. It would overwrite people's commits (it can be visible on the conflict scenario a bit).

But thinking about this more, the way we are creating the diff, it sort of contains the xcopy inside just we leave there room for conflicts with the target branch. I can elaborate on this further as I understand this is a very vague description but an xcopy version of the document can be seen here for comparison:
https://github.com/premun/arcade/blob/0604de9a2789c8518b16bbe8f73db1bea2460994/Documentation/UnifiedBuild/Full-Code-Flow.md#the-codeflow-algorithm

I am still very much open to a change of heart here as I had that already but after laying it out, I believe that even the current proposal without xcopy doesn't allow for an accumulation of delta.

premun commented:
cc @agocke as you expressed interest in the code flow algorithm at the last team-wide sync if I am not wrong? Happy to have as much feedback as possible.

jkotas commented on Documentation/UnifiedBuild/Full-Code-Flow.md:
```diff
@@ -0,0 +1,414 @@
+
+## The code flow algorithm
+
+This section describes the details of moving the code between product repositories and the VMR. The algorithm will work differently in each direction to achieve maximum fleuncy and minimize the amount of conflicts developers need to tend to but also ensure that conflicting changes manifest as conflicts and changes are not overridden without a trace.
```
```suggestion
This section describes the details of moving the code between product repositories and the VMR. The algorithm will work differently in each direction to achieve maximum fluency and minimize the amount of conflicts developers need to tend to but also ensure that conflicting changes manifest as conflicts and changes are not overridden without a trace.
```

jkotas commented on Documentation/UnifiedBuild/Full-Code-Flow.md:
```diff
@@ -0,0 +1,414 @@
+The numbered steps are described in more detail below:
+
+1. This is the current normal process for making a change to an individual repository. Nothing changes.
+2. Currently, each official build of each repo publishes itself via darc which registers the commit and the set of built packages into the BAR and is assigned to zero or more channels. There is a lot of configuration effort in which repositories publish from which branches to which channels. We intend to keep this existing setup in place and piggy back on this. The only change to the current stat ...
```
Official build is complicated moving part that takes a long time. The whole point of VMR is to have just one and only official build in the VMR repo. We should need to have full dotnet/runtime official build to trigger the forward flow.

If you want to piggyback on maestro for this, it is ok - but the triggered should be a dummy official build that publishes a trivial dummy package as semaphore.

premun replied:
We had this idea of a no-op official builds that would just ping the system and I think we would eventually converge there but I am not sure at the moment we want that right off the bat.
From the Vertical Build discussions, it seems that repositories might be building fewer things in the VMR vertical build than what the repo build does at the moment. It seems some directly-non-related-to-product build outputs (e.g. packages consumed by other repos which are not part of the SDK) won't get built/published during the VMR vertical build? I can then see a room for the current Maestro to keep flowing these from the product repo to wherever while the product packages would be dealt with through the VMR?

Thoughts? @mmitche?

jkotas replied:
The multi-product repos like dotnet/roslyn will need to keep their current official builds to product their other deliverables. Single-product repos like dotnet/runtime do not need that.

Special care needs to be taken to avoid package version conflicts between repo-specific official builds and the VMR build. If both of these builds produce same package, they need to them to different feeds or use some other scheme to avoid version conflicts. The easiest way to deal with this is to say that packages produced by VMR should not be published by VMR build.

The official builds in dotnet/runtime require constant attention. They fail for all sort of reasons. I think setting up the no-op official build for dotnet/runtime and other large repos has to be part of the plan. Otherwise, we will double the amount of effort required to keep the system running.

premun replied:
@steveisok fyi, I believe this might be important for the VB controls so that we can produce distinct set of assets based on the context

mmitche replied:
I filed an issue on this in source-build.

jkotas commented on Documentation/UnifiedBuild/Full-Code-Flow.md:
```diff
@@ -0,0 +1,414 @@
+    runtime--1. A change is merged into dotnet/runtime,\nmirrored to AzDO and the official build starts-->runtimeCI
+    runtimeCI--2. Build is added to the .NET 9 channel,\nsubscription from runtime to dotnet/dotnet triggered-->maestro
+    maestro--3. Maestro notices a VMR subscription\ntriggered, calls the backflow service-->backflow
+    backflow--4. A branch with the appropriate code\nchanges in dotnet/dotnet is created-->maestro
```
It is a bit confusing that backflow service is responsible for forward flow. Should it have a different name? E.g. codeflow service?

premun replied:
You are right, I should change it in this doc, it seems that the new service will be called "Conductor" as it relates to "Maestro" in theme.

Eventually, all functionality from Maestro will be migrated over to the new service and current Maestro killed. This is due to Maestro running on Service Fabric which has some engineering drawbacks. So eventually we would have just one service and that service will be called something like Conductor but it should encompass the full functionality of the current and new systems.

Other ideas were:
- Yoyo
- Baton

Feel free to weigh in if you have an idea and care enough! :)

jkotas commented on Documentation/UnifiedBuild/Full-Code-Flow.md:
```diff
@@ -0,0 +1,414 @@
+### Configuration location
+
+When changes are made to installer's `source-mappings.json`, the first next synchronization takes them into account already.
+This is a very handy mechanism as we can react to incoming changes of every individual repository right away. For instance, if a new commit of a repository brings in binaries, we can add the cloaking rules in the very same commit and the binaries will never reach the VMR.
```
The current policies associated with source-mapping.json are very source build specific. 

I expect that we will end up with files and components in the VMR to make Microsoft build works, but we do not want to have the same files or components in the source build tarball consumed by the distro.

Thoughts?

mmitche replied:
@jkotas Microsoft build will definitely use components and files that are not available to RedHat, but the source layouts that we build will be the same. This is an explicit goal of UB. We want to be able to hand them the sha that we build, and they build it. The behavior may be somewhat different (some binary packages wouldn't be restored, for instance), but the source layout would be the same. Note that aside from a few repos that are not in source-build because they don't have to be, this is the way things work today.

I don't think maintaining a permanent delta is a good idea. I think it's a maintenance and codeflow minefield, a bad look for transparency, and I think that most of the places where a permanent delta might be a solution are actually just cases where we should flow binaries into the product when building as Microsoft. Cases where we need Microsoft specific code in a VMR are probably bad code smells more than anything else. To require a permanent delta, we would need to have a situation where we have build of public sources flowing into private sources, which then flows back into a public sources. I don't believe we've ever had this show up in the past. Components that are closed source do show up (e.g. wpf-int), but they are always leaves. We will continue to support such workflows with the VMR.

jkotas replied:
As an example, take the following excludes from https://github.com/dotnet/installer/blob/main/src/VirtualMonoRepo/source-mappings.json#L70-L71 :

```
                "src/Installers/Windows/AspNetCoreModule-Setup/*",
                "src/SignalR/clients/java/signalr/gradle/wrapper/gradle-wrapper.jar"
```

These excludes were required for source build (source build partners do not like binaries and files with problematic licenses in the tarball). I expect that these excludes are going to break Microsoft build.

What is going to happen with these excludes in VMR model?

mmitche replied:
I think we have to remove the excludes. Binaries are perhaps easier to deal with, since we can pull them in as packages. Non-compliant licenses are tougher.

If we allow a permanent delta though, we have to maintain two equivalent source bases and a relationship between them. And I think this would be a takeback from where we are today. We tag our repos based on what we build, and RH/Canonical/other partners build off those shas.

If it does become problematic to remove the excludes, I could see us creating tools to take our release shas and create a delta branch based on a set of excludes, but I'd like to avoid that.

/cc @MichaelSimons

premun replied:
Needless to say, the exclusions are actually making the implementation of the algorithm proposed here much more difficult.
It will also mean all the operations like creating the PR branches take much longer due to the way we will need to create the commits.

So if we could scratch exclusions, we should make the decision early.

premun replied:
Nevermind, I made everything work with exclusions and we will need them anyway for our own synchronization reasons.

jkotas commented on Documentation/UnifiedBuild/Full-Code-Flow.md:
```diff
@@ -0,0 +1,414 @@
+
+You can notice several features:
+- No (git) conflicts appear. This is because this concrete example considers a single file that is chronologically changed from `one` to `five` in gradual steps. In such case, we should not expect any conflicts and this is by design. In cases where most of the changes happen in the individual repository, we expect the code to flow fluently.
+- The whole flow is comparable to a dev working in a dev branch within a single repository - the dev branch being the VMR where the dev merges the main branch in between the work (this is the forward flow). The dev then opens a PR against the main branch (the repository in this case). Wherever there would be conflicts in a single repository case, we would get conflicts here too and this is by des ...
```
> The whole flow is comparable to a dev working in a dev branch within a single repository

I like this explanation.

jkotas commented on Documentation/UnifiedBuild/Full-Code-Flow.md:
```diff
@@ -0,0 +1,414 @@
+
+## The code flow algorithm
+
+This section describes the details of moving the code between product repositories and the VMR. The algorithm will work differently in each direction to achieve maximum fleuncy and minimize the amount of conflicts developers need to tend to but also ensure that conflicting changes manifest as conflicts and changes are not overridden without a trace.
```
The algorithm should be explained as a pseudo code.

The picture is useful to demonstrate how it works on an examples, but it is very ambiguous description.

premun replied:
@jkotas I have added pseudocode but am not sure if I chose the right level of detail. Please let me know if it's readable this way.

I might also try to work a bit more on some confirmations that the xcopy won't overwrite people's changes blindly. But there's many situations to consider so I need to choose one.

jkotas commented:
> But thinking about this more, the way we are creating the diff, it sort of contains the xcopy inside just we leave there room for conflicts with the target branch.

I am not able to tell from the picture.

This is what my xcopy-based version of the algorithm would be.

## Forward flow

1. Go through dotnet/dotnet history and find the last commit titled "Codeflow from dotnet/runtime"
2. `git checkout` commit found in step 1
3. xcopy sources from dotnet/runtime over dotnet/dotnet, with source mappings applied
4. `git commit` (w/ any new files added/removed)
5. Submit PR. We assume that the PR validation is going to rebase this delta against the current main that will surface any conflicts.

## Backward flow:

1. Go through dotnet/runtime history and find the last commit titled "Codeflow from dotnet/dotnet"
2. `git checkout` commit found in step 1
3. xcopy sources from dotnet/dotnet over dotnet/runtime
4. `git commit` (w/ any new files added/removed)
5. Submit PR. We assume that the PR validation is going to rebase this delta against the current main that will surface any conflicts.

jkotas commented:
> This is what my xcopy-based version of the algorithm would be
> Backward flow

If you prefer to have the backward direction done differently like what you have suggested in the doc, it is fine with me. The property that I am looking for is that one of the directions is going to self-heal any accidental diffs.

premun commented:
> If you prefer to have the backward direction done differently like what you have suggested in the doc, it is fine with me. The property that I am looking for is that one of the directions is going to self-heal any accidental diffs.

Understood. I will identify the exact places where I think we get this property and highlight those in the doc.

The other property that I am hoping to get out of this, and that I believe is sort of equitable to yours, is this:
> If things go messy and we don't understand why certain changes/conflicts appear in the flow, we should not need to go back several flows to figure out where they come from, but the most recent inflow should draw an enough clear synchronization line between the repos so that it can be considered an equal state. From there we should be able to backtrack changes easily.

I believe without this property, we will lose all sense fast. An xcopy gives you such nice behaviour out-of-the-box but I think there's a cost to it. I will prepare a diagram where I believe xcopy deletes changes without a trace and I will try to describe why I think the proposed algorithm still gives us this clear synchronization line which we do not need go behind really.

agocke commented on Documentation/UnifiedBuild/Full-Code-Flow.md:
```diff
@@ -0,0 +1,414 @@
+`1` and `2` denote some previous synchronization point.
+`3` Commit in the repository changes the contents of `A.txt` to `two`.  
+`4` Commit in the repository changes the contents of `A.txt` to `three`. Forward flow also starts at that point (this is arbitrary).  
+`5` A forward-flow branch (green) is created in the VMR. The branch is based on the commit of last synchronization (`2`). The content on the commit is [described below](#forward-flow-commit). A PR from this branch is opened.  
```
I can't seem to find this link

premun replied:
Fixed

agocke commented on Documentation/UnifiedBuild/Full-Code-Flow.md:
```diff
@@ -0,0 +1,414 @@
+`10` A backflow branch (blue) is created in the repository. The branch is based on the commit of last synchronization's (`2`) base commit. The content on the commit is [described below](#backflow-commit). A PR from this branch is opened.  
+`11` A commit is made to the main branch of the repository, content is unrelated.  
+`12` A commit is made to the main branch of the repository, content is unrelated.  
+`13` The PR is merged, effectively updating `A.txt` from `three` to `five`.
```
How did we get to five? The last commit mentioned is four.

premun replied:
The VMR commit `9` changes the file `four` from `five`. It's the wrong colour on the diagram and the text describing `9` above is wrong, will fix. The outcome is correct though. Thanks for noticing.

agocke commented on Documentation/UnifiedBuild/Full-Code-Flow.md:
```diff
@@ -0,0 +1,414 @@
+The delta between the repositories when commit `9` (which we will backflow) happens is technically commits `6`, `7` and `9` is visualized as the purple diff between `9` and `5`. This diff correctly represents the delta because:
+- It contains the last known snapshot of the repository (`5`)
+- All commits that happened in the VMR in the meantime (since the last commit) - the commit `7`.
+- The other commits that happened in the VMR since the sync `9`.
```
Something that it seems like it's worth calling out: not running tests in the VMR means that we're dependent on backflow for validation. Continuing with forward flow while backflow is pending means that we could be moving into invalid states very quickly.

mmitche replied:
This is sort of the case today actually. You can have dotnet/runtime build 10 times and flow into dotnet/sdk, but be broken at dotnet/aspnetcore. I think likely to continue seeing those cases, but I don't get the sense that we would be hitting that at higher rates.

However, it is important to note that we are reliant on backflow for product validation.

agocke commented:
\+ @jtschuster for background info

mmitche commented on Documentation/UnifiedBuild/Full-Code-Flow.md:
```diff
@@ -0,0 +1,414 @@
+    Maestro--If source repo is arcade-->EngCommon
+
+    EngCommon-->VersionFiles
+    Maestro--If source repo is VMR-->CallBackflowService
```
It is possible for the source repo to be arcade, and the target repo to be the VMR.

mmitche commented on Documentation/UnifiedBuild/Full-Code-Flow.md:
```diff
@@ -0,0 +1,414 @@
+
+### Last flow detection
+
+For the above to work correctly, we need to be able to tell which situation we're in and which direction the last flow happened. For this to work, we need to store the last SHA of the counterpart repository that we have synchronized from the last. In the present VMR, this information is already present in the `source-manifest.json` file. For flowing from the VMR into an individual repository, we  ...
```
I think we should add a separate section to V.D.xml for the VMR->repo flow. We will have VMR shas in the dependency details. However, in servicing, they will start to vary because of incremental servicing.

premun replied:
You mean if we only service in the VMR with no flow on?

premun replied:
Should we also put all of the configuration in `V.D.xml` instead of the proposed `vmr-synchronization.json` file?

premun commented:
@jkotas thanks for the feedback. You correctly found some typos and the code was kind of messy. I have refactored the pseudocode to be simpler - it has one entrypoint which decides the direction and I changed the last flow detection abstraction to also read better. Sorry for changing this so much but I think it's an improvement.

I have also implemented parts of this to validate it better with some E2E tests so the last flow abstraction can be seen here for example:
https://github.com/premun/arcade-services/blob/452933457e687203c0c076d1cfd0175ad5c97fc2/src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrBackflowManager.cs#L280-L309

premun commented on Documentation/UnifiedBuild/VMR-Full-Code-Flow.md:
```diff
@@ -0,0 +1,558 @@
+    # We recreate the last flown state, apply new diff on top and create a PR
+    # The changes that were already merged before (the previously flown state)
+    # will be transparently hidden when resolving the conflict in the new PR
+    flow($last_flow.source_sha, $source_repo, $target_repo)
```
Fixed
